### DISCUSSION TITLE
YaSolR Hysteresis and/or average setup for classic relays load

### DISCUSSION THREAD

**Discussion by hallard** [mathieucarbou/YaSolR#37] [Ideas]
When using relay1 and relay2 (not the PWM output  or whatsoever) just to drive some loads on device where PWM is not possible (electronic Heaters or Heating Pump for example) I was thinking it could be great to have some Hysteresis or dual setup to avoid flooding ON/OFF relay when production varies close to setup. Let's example:

- setup relay to `ON` above 700W but goes `OFF` under 500W 
- also a possibility to average solar production for example to 5 last minutes and use this averaged values to drive relay, this would for example reduce above hysterisis setup to 700W and 650W because averaged value has slower moves.

What do you think?

**Comment by mathieucarbou** [comment]
Yeah I knew this question would come .

Relays in YaSolR are not designed to work like the F1ATB for example, to control any kind of load.

They were designed to complement the routing and offer a way to add steps to a water tank having for example 3 heating elements, in order to reduce the nominal power going through the triac or SSR.

They could be used also on other external resistive and constant loads.

The problem with external loads is that the router cannot correctly measure their consumption / resistance value and also cannot know whether the load is constant or variable (like a pump). Consequently, there is no way to accurately know  when to stop the load. Hysteresis/ averaging for that use case will just make the routing worse because there will be a lot of import and export.

So I decided to let this feature to the user and instead focus YaSolR on what it should fo best: ensure a correct and reliable routing.

YaSolR provides all the API necessary (rest, mqtt, ha intégration). It exposes the routed power, grid power and virtual grid power, which is the grid power as it would be if the router was not routing. Thanks to this value, the user can control through HA or Shelly's his own on/off and hystereis depending on the equipment type, virtual grid power, etc. But the real issue is to determine when to switch the load and when to stop. For example, a pump starting takes a lot of power, then consumes less. What the user wants ? Start when enough excess to cover the load spike of the pump or only it's power consumption when running ? And at which pace ? Some systems with pumps can have several speeds...

HA or Shelly's are better to handle this kind of automations.

So I didn't want to overcomplicate YaSolR with such feature because this is not reliable as explained above, which would highly impact the routing accuracy, even on a constant load because of the hystesis/averaging .

If there is a way / design to have this feature work in a reliable way, and be able to correctly determine the power consumption of the external elements to still provide a good routing accuracy, I will consider, but for now I found no way to correctly ensure a good routing with such feature.

**Reply by mathieucarbou** [reply]
What could be implemented though, is a way to "activate" a load when the excess is above a level, and keep it active for a certain duration, regardless of the Gris import/export.

The relays would just act as a switch on feature that will allow the start-up of a load at a given threshold.

And eventually if set, will stop the load after some time..

Such design won't impact the routing, and there is no need to know the connected loads

**Reply by hallard** [reply]
Fair enough completely understand your point, forgot to think that you need routed power in device for calculation. 
I was more thinking just as you mention, an ON/OFF feature whatever is connected to and in this case no use of other routing feature of YaSolR of course.

I'm doing lot of things on the other side (HA/NodeRED/,...) so I have no issue implementing this feature and correct, I have in mind to drive Shelly Plug where my Ecoflow Delta 2 Max is connected to.
I have also in mind to setup charging power of Delta2 with Ecoflow API, like that, 200W over production, 200W Charging, 300W, 300W Charging and so on, basic, but better than nothing and injecting for free.

**Reply by mathieucarbou** [reply]
Yes, most of the time,  people have more complex use cases like that than just being able to switch a pump.

I added though a note to just implement an ON feature with an optional timeout. Would be equivalent as saying: if there is 100W of excess, then start the pump for 1 hour.

I will see how relevant it is. Because on a cloudy / shady day, it could just not work at all.

**Reply by hallard** [reply]
yeah may be not so relevant, not sure I will use it but could help, setting time for 5 min could be better, or worse depending on what's connected :-)
